File: 03.cpp
#include <iostream>

long LargestPrimeFactor(long long n)
{
	long long start = 2;
	while (n > start*2)
	{
		if (n%start == 0)
			n = n/start;
		else
			start++;
	}
	return n;
}


int main (int argc, char ** argv)
{
	long long n = strtoll(argv[1], NULL, 10);
	std::cout << n << std::endl;
	std::cout << LargestPrimeFactor(n) << std::endl;
}

File: 03.hs
-- works!
-- Find the largest prime factor of 600851475143.

input = 600851475143
largestPrimeFactor n = lpf n 2
  where
    lpf n start
      | n <= (start * 2) = n
      | mod n start == 0 = lpf ( div n start) start
      | otherwise = lpf n (start+1)

